navcoin/navcoin-core
# Fix for verifychain

Fixes #630

See issue for more details.

### Changes introduced

- New RPC command `getcfunddbstatehash` covered by functional test cfunddb-statehash.py
- New state for payment requests `6` when those are paid.
- The payment request parameter `paidOnBlock` is substituted by `stateChangedOnBlock` when state is `6` 
- New structure for storing the state history of CFundDB entries. Those are stored in a map associating blockhash and state, allowing to directly revert state transitions when reorganizations are seen.
- `verifychain` now checks for the consistency of the CFundDB state hash when level `4` is specified.

## Important

This set of changes will require older clients to reindex on launch, keeping the node offline for some hours at best. In order to reduce downtime, node operators can proceed as follows if needed:

- Close node with old version.
- `mkdir /tmp/reindexdata`; `cp -rf <data_folder>/blocks /tmp/reindexdata/`; `cp -rf <data_folder>/chainstate /tmp/reindexdata/` 
- Reopen node with old version. It will be again online
- Launch in parallel a second instance of the node, this time using the new version with the parameters `-reindex -datadir=/tmp/reindexdata/`
- Once the reindex finishes, close both nodes and copy back the reindexed data.
- `rm -rf <data_folder>/blocks <data_folder>/chainstate`; `cp -rf /tmp/reindexdata/* <data_folder>`
- Relaunch new version of the node.

### What to test

- Launch a wallet with an old copy of the blockchain. It should show that the database is corrupted and that reindex is needed. Test both gui and headless
- Launch a fully synced wallet and execute `verifychain 4 0`. This should return true.
- Partial runs of `verifychain 4 n` return true (being n an arbitrary value between 1 and the length of the chain).
- Wallet should be able to sync correctly
